### PR TITLE
[WC-1699] Selection helper widget

### PR DESCRIPTION
**Pull request by r0b1n**
### Pull request type

New feature (non-breaking change which adds functionality)

---

### Description

- Add selection context to PW commons.
- Provide selection context to header widgets in Gallery widget
- Add Selection Helper widget that could be placed in the header of gallery

### What should be covered while testing?

- Selection Helper widget works with Gallery widget. 
- Gallery widget allows place widget Selection Helper when Multi selection is enabled (header is shown while no sorting and filtering are enabled)
- Filtering and sorting widgets still work

### CODE REVIEW

File: packages/pluggableWidgets/selection-helper-web/prettier.config.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+const base = require("@mendix/pluggable-widgets-tools/configs/prettier.base.json");
+
+module.exports = {
+    ...base,
+    plugins: [require.resolve("@prettier/plugin-xml")]
+};
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/prettier.config.js:6]: You on purpose use base from PWT?
I would recommend use our current config that we use for other widgets
```sh
pnpm add -D "@mendix/prettier-config-web-widgets@*"
```

```js
module.exports = require("@mendix/prettier-config-web-widgets");
```

File: packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorConfig.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+//     // Customize your pluggable widget appearance for Studio Pro.
+//     return {
+//         type: "Container",
+//         children: []
+//     }
+// }
+
+// export function getCustomCaption(values: SelectionHelperPreviewProps, platform: Platform): string {
+//     return "SelectionHelper";
+// }
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorConfig.ts:42]: Why is this commented?

File: packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorConfig.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { hidePropertiesIn, Properties } from "@mendix/pluggable-widgets-tools";
+import { SelectionHelperPreviewProps } from "../typings/SelectionHelperProps";
+
+export function getProperties(
+    values: SelectionHelperPreviewProps,
+    defaultProperties: Properties /* , target: Platform*/
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorConfig.ts:6]: Let's remove commented arg

File: packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorPreview.tsx [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { ReactElement, createElement } from "react";
+import { SelectionHelperPreviewProps } from "../typings/SelectionHelperProps";
+
+export function preview(props: SelectionHelperPreviewProps): ReactElement {
+    return <div>{JSON.stringify(props)}</div>;
+}
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.editorPreview.tsx:6]: Ah, okay, there is WIP

File: packages/pluggableWidgets/selection-helper-web/src/components/CheckBox.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        <input
+            id={id}
+            type="checkbox"
+            className="three-state-checkbox"
+            ref={checkboxRef}
+            checked={value !== "none"}
+            onChange={onChange}
+        />
+    );
+}
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/src/components/CheckBox.tsx:33]: Nice solution!

File: packages/pluggableWidgets/selection-helper-web/src/components/CheckBox.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { createElement, ReactElement, useEffect, useRef } from "react";
+import "../ui/CheckBox.scss";
+
+interface CheckBoxProps {
```
[the remark this answers is not in the document]
Comment (r0b1n) [reply packages/pluggableWidgets/selection-helper-web/src/components/CheckBox.tsx:4]: We don't pass anything really meaningful at the moment, I don't think this is needed at the moment

File: packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.xml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<widget id="com.mendix.widget.web.selectionhelper.SelectionHelper" pluginWidget="true" needsEntityContext="true" offlineCapable="true" supportedPlatform="Web" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
+    <name>Selection Helper</name>
+    <description>My widget description</description>
```
Comment (iobuhov) [inline packages/pluggableWidgets/selection-helper-web/src/SelectionHelper.xml:4]: I think we need a better description 😂

**Review by iobuhov — CHANGES_REQUESTED** [review]

**Review by iobuhov — DISMISSED** [review]

**Review by iobuhov — APPROVED** [review]

**Review by leonardomendix — APPROVED** [review]

**Pull request merged**